### PR TITLE
swap

### PR DESCRIPTION
**Pull request by mhhf**
swap swap

### CODE REVIEW

File: src/uniswap.act.md [hunk 1]
```diff
@@ -734,6 +734,100 @@ calls
...
+    BlockTimestampLast : uint32
+
+storage
+
+    reserve0_reserve1_blockTimestampLast |-> #WordPackUInt112UInt112UInt32(Reserve0, Reserve1, BlockTimestampLast) => #WordPackUInt112UInt112UInt32(Balance0 - amount0Out, Balance1 - amount1Out, TIME)
+    token0   |-> Token0
+    token1   |-> Token1
+    lockState |-> 1
+    price0CumulativeLast |-> Price0 => #if TIME - BlockTimestampLast =/= 0 #then ((((pow112 * Reserve1) / Reserve0) * (TIME - BlockTimestampLast)) + Price0) #else Price0 #fi
+    price1CumulativeLast |-> Price1 => #if TIME - BlockTimestampLast =/= 0 #then ((((pow112 * Reserve0) / Reserve1) * (TIME - BlockTimestampLast)) + Price1) #else Price1 #fi
```
Comment (xwvvvvwx) [inline src/uniswap.act.md:765]: I think you're missing the checks for non zero reserves before updating the price accumulators here

File: src/uniswap.act.md [hunk 2]
```diff
@@ -734,6 +734,100 @@ calls
...
+where
+
+    Amount0In := (Reserve0 - amount0Out) < (Balance0 - amount0Out)
+    Amount1In := (Reserve1 - amount1Out) < (Balance1 - amount1Out)
+    K00 := ((Reserve0 * Reserve1) * 1000000) <=  (((Balance0 - amount0Out) * 1000) - ((Balance0 - (amount0Out + (Reserve0 - amount0Out))) * 3)) * (((Balance1 - amount1Out) * 1000) - ((Balance1 - (amount1Out + (Reserve1 - amount1Out))) * 3))
+    K01 := ((Reserve0 * Reserve1) * 1000000) <= ((((Balance0 - amount0Out) * 1000) - ((Balance0 - (amount0Out + (Reserve0 - amount0Out))) * 3)) * ((Balance1 - amount1Out) * 1000))
+    K10 := ((Reserve0 * Reserve1) * 1000000) <= (((Balance0 - amount0Out) * 1000) * (((Balance1 - amount1Out) * 1000) - ((Balance1 - (amount1Out + (Reserve1 - amount1Out))) * 3)))
+
+
+iff
```
Comment (xwvvvvwx) [inline src/uniswap.act.md:786]: I think we should revert if `lockState =/= 1` here?

File: src/uniswap.act.md [hunk 3]
```diff
@@ -734,6 +734,100 @@ calls
...
+
+    VCallValue == 0
+    VCallDepth < 1024
+    amount0Out < Reserve0
+    amount1Out < Reserve1
+    amount0Out =/=K 0 orBool amount1Out =/=K 0
+    Amount0In orBool Amount1In
+    (notBool Amount0In) orBool (notBool Amount1In) orBool K00
```
Comment (xwvvvvwx) [inline src/uniswap.act.md:794]: Maybe I'm just not so quick in logic, but I'm having a hard time reading these conditions. I would find `(Amount0In and Amoun1In) impliesBool K00` easier to parse. I had to go write out the truth tables to satisfy myself that it's equivalent.

File: src/uniswap.act.md [hunk 4]
```diff
@@ -734,6 +734,100 @@ calls
...
+
+if
+
+    Token0 =/= Token1
+    Token0 =/= to
+    Token1 =/= to
+    ACCT_ID =/= to
+    calldata ==K .WordStack
+    TIME < pow32
+    TIME >= BlockTimestampLast
```
[the remark this answers is not in the document]
Comment (MrChico) [reply src/uniswap.act.md:828]: This is probably better to put as an `if` assumption

File: src/uniswap.act.md [hunk 5]
```diff
@@ -734,6 +734,100 @@ calls
...
+    ((pow112 * Reserve0) / Reserve1) * (TIME - BlockTimestampLast)
+    (((pow112 * Reserve1) / Reserve0) * (TIME - BlockTimestampLast)) + Price0
+    (((pow112 * Reserve0) / Reserve1) * (TIME - BlockTimestampLast)) + Price1
+
+
+if
+
+    Token0 =/= Token1
+    Token0 =/= to
+    Token1 =/= to
```
[the remark this answers is not in the document]
Comment (mhhf) [reply src/uniswap.act.md:824]: ah, you're right! will fix it in the new update

File: src/uniswap.act.md [hunk 6]
```diff
@@ -734,6 +734,100 @@ calls
...
+    Balance0 - (amount0Out + (Reserve0 - amount0Out))
+    Balance1 * 1000
+    (Balance1 - Reserve1) * 3
+    (Balance0 - amount0Out) * 1000
+    (((Balance0 * 1000) * (((Balance1 - amount1Out) * 1000) - ((Balance1 - (amount1Out + (Reserve1 - amount1Out))) * 3))))
+    (((Balance0 - amount0Out) * 1000) - ((Balance0 - (amount0Out + (Reserve0 - amount0Out))) * 3)) * (((Balance1 - amount1Out) * 1000) - ((Balance1 - (amount1Out + (Reserve1 - amount1Out))) * 3))
+    ((pow112 * Reserve1) / Reserve0) * (TIME - BlockTimestampLast)
+    ((pow112 * Reserve0) / Reserve1) * (TIME - BlockTimestampLast)
+    (((pow112 * Reserve1) / Reserve0) * (TIME - BlockTimestampLast)) + Price0
+    (((pow112 * Reserve0) / Reserve1) * (TIME - BlockTimestampLast)) + Price1
```
[the remark this answers is not in the document]
Comment (mhhf) [reply src/uniswap.act.md:817]: hmm, will try it out

File: src/uniswap.act.md (hunk 2 above)
Comment (mhhf) [reply src/uniswap.act.md:786]: yes! will fix it!

File: src/uniswap.act.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (mhhf) [reply src/uniswap.act.md:828]: can try this, but fear this will be clashing with packed storage

File: src/uniswap.act.md (hunk 1 above)
Comment (mhhf) [reply src/uniswap.act.md:765]: actually i think its not necessary, its implied from the other assumptions and the behaviour

File: src/uniswap.act.md (hunk 3 above)
Comment (mhhf) [reply src/uniswap.act.md:794]: ah, didn't know implies Bool is a thing, tho i actually prefer this form. here you see that either one has to be true

File: src/uniswap.act.md (hunk 1 above)
Comment (mhhf) [reply src/uniswap.act.md:765]: but maybe i will need to add this with allowing some stuff to overflow where the nonzero constraints came from, i will make a run and report back

File: src/uniswap.act.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (xwvvvvwx) [reply src/uniswap.act.md:828]: Also fine leaving this in the `if` block as we discussed in the call this morning.

File: src/uniswap.act.md (hunk 1 above)
Comment (xwvvvvwx) [reply src/uniswap.act.md:765]: You can check the `sync` and `mint` specs for examples of specs that deal with overflow in `_update`.

File: src/uniswap.act.md (hunk 3 above)
Comment (xwvvvvwx) [reply src/uniswap.act.md:794]: That's a good point. I'm fine to leave this as is.

**Pull request merged**